src/Shared/StackTrace/StackFrame/PortablePdbReader.cs

```diff
@@ -66,60 +66,77 @@ public void PopulateStackFrame(StackFrameInfo frameInfo, MethodBase method, int
 
         private MetadataReader GetMetadataReader(string assemblyPath)
         {
-            MetadataReaderProvider provider = null;
-            if (!_cache.TryGetValue(assemblyPath, out provider))
+            if (string.IsNullOrEmpty(assemblyPath))
+            {
+                return null;
+            }
+
+            if (_cache.TryGetValue(assemblyPath, out var provider))
+            {
+                return provider.GetMetadataReader();
+            }
+
+            if (TryGetPdbPath(assemblyPath, out var pdbPath) && File.Exists(pdbPath))
             {
-                var pdbPath = GetPdbPath(assemblyPath);
+                var pdbStream = File.OpenRead(pdbPath);
 
-                if (!string.IsNullOrEmpty(pdbPath) && File.Exists(pdbPath) && IsPortable(pdbPath))
+                if (IsPortable(pdbStream))
                 {
-                    var pdbStream = File.OpenRead(pdbPath);
                     provider = MetadataReaderProvider.FromPortablePdbStream(pdbStream);
-                }
 
-                _cache[assemblyPath] = provider;
+                    _cache[assemblyPath] = provider;
+
+                    return provider.GetMetadataReader();
+                }
             }
 
-            return provider?.GetMetadataReader();
+            return null;
         }
 
-        private static string GetPdbPath(string assemblyPath)
+        private static bool TryGetPdbPath(string assemblyPath, out string pdbPath)
         {
-            if (string.IsNullOrEmpty(assemblyPath))
-            {
-                return null;
-            }
-
             if (File.Exists(assemblyPath))
             {
                 var peStream = File.OpenRead(assemblyPath);
 
                 using (var peReader = new PEReader(peStream))
                 {
-                    foreach (var entry in peReader.ReadDebugDirectory())
+                    try
                     {
-                        if (entry.Type == DebugDirectoryEntryType.CodeView)
+                        foreach (var entry in peReader.ReadDebugDirectory())
                         {
-                            var codeViewData = peReader.ReadCodeViewDebugDirectoryData(entry);
-                            var peDirectory = Path.GetDirectoryName(assemblyPath);
-                            return Path.Combine(peDirectory, Path.GetFileName(codeViewData.Path));
+                            if (entry.Type == DebugDirectoryEntryType.CodeView)
+                            {
+                                var codeViewData = peReader.ReadCodeViewDebugDirectoryData(entry);
+                                var peDirectory = Path.GetDirectoryName(assemblyPath);
+                                pdbPath = Path.Combine(peDirectory, Path.GetFileName(codeViewData.Path));
+                                return true;
+                            }
                         }
                     }
+                    catch
```

Comment:
Should the exception still be logged at a debug level?

Reply:
Yeah, I thought about that. Might be valuable. But is it OK to take a dependency on `ILogger` or something in these shared source packages?

Reply:
Try it and we'll see if we hit any conflicts.

Reply:
We shouldn't have a dependency on `ILogger` at this level of component.

Reply:
It'd look a bit like this; https://github.com/aspnet/AspNetCore/commit/90c65b3482cd42907c213de76aa7b0e579e801fb

Reply:
@davidfowl what's wrong with optionally passing in the caller's logger?

Reply:
@davidfowl what's wrong with optionally passing in the caller's logger?

Reply:
The caller should just catch the exception.

Reply:
The `ExceptionDetailsProvider` already has a dependency on Microsoft.Extensions.FileProviders though.

Reply:
That's the component that should catch the exception then.

Reply:
So what's the verdict? Move the `try`/`catch` up into `ExceptionDetailsProvider`?

Reply:
Yep

Reply:
There's an issue with moving it higher up in the stack. The call that would throw is

https://github.com/aspnet/AspNetCore/blob/8e05fb0cf450afdf629b275435c69c836ae8df68/src/Shared/StackTrace/StackFrame/StackTraceHelper.cs#L61

If the exception isn't caught inside the `PopulateStackFrame` call, there's no way to get `StackFrameInfo` for other frames in the stack, which could be valuable. You'd basically end up with no info.

Reply:
I'm missing something, why would you miss all frames if we failed to populate info for this single frame?

Reply:
Because the exception would bubble up through `StackTraceHelper.GetFrames` and be caught in `ExceptionDetailsProvider.GetDetails`. At that point, it would have no choice but to use `Enumerable.Empty<StackFrameInfo>` for that exception.

Reply:
Oh I was assuming you would still populate remaining frames. You need an out parameter from that method with an aggregate exception.